web-api/src/mind.rs

```diff
@@ -17,31 +17,116 @@
 use std::{collections::HashMap, fs::File, path::Path};
 
 use csv::{DeserializeRecordsIntoIter, Reader, ReaderBuilder};
+use itertools::Itertools;
 use ndarray::{Array, ArrayView};
-use rand::{seq::SliceRandom, thread_rng};
-use serde::Deserialize;
+use rand::{
+    seq::{IteratorRandom, SliceRandom},
+    thread_rng,
+};
+use serde::{de, Deserialize, Deserializer};
 use xayn_ai_coi::nan_safe_f32_cmp_desc;
 
+use crate::models::DocumentId;
+
+#[derive(Debug, Deserialize)]
+struct ViewedDocument {
+    document_id: DocumentId,
+    was_clicked: bool,
+}
+
+impl ViewedDocument {
+    fn new(document_id: DocumentId, was_clicked: bool) -> Self {
+        Self {
+            document_id,
+            was_clicked,
+        }
+    }
+}
+
+fn deserialize_viewed_documents<'de, D>(deserializer: D) -> Result<Vec<ViewedDocument>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let s = String::deserialize(deserializer)?;
+    let viewed_documents = s.split(' ');
+    viewed_documents
+        .map(|viewed_document| -> Result<_, D::Error> {
+            let mut parts = viewed_document.split('-');
+            let document_id = parts
+                .next()
+                .ok_or_else(|| de::Error::custom("missing document id"))?;
+            let document_id = DocumentId::new(document_id).map_err(de::Error::custom)?;
+            let was_clicked = parts
+                .next()
+                .ok_or_else(|| de::Error::custom("missing was_clicked"))?;
+            let was_clicked = match was_clicked {
+                "0" => Ok(false),
+                "1" => Ok(true),
+                _ => Err(de::Error::custom("invalid was_clicked")),
+            }?;
+            Ok(ViewedDocument::new(document_id, was_clicked))
+        })
+        .collect()
+}
+
+fn deserialize_clicked_documents<'de, D>(deserializer: D) -> Result<Vec<DocumentId>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let s = String::deserialize(deserializer)?;
+    let viewed_documents = s.split(' ');
+    viewed_documents
+        .map(|document| -> Result<_, D::Error> {
+            let document_id = DocumentId::new(document).map_err(de::Error::custom)?;
+            Ok(document_id)
+        })
+        .collect()
```

Comment:
:art: we can write this more concise for better readability. the deserialization, iteration and mapping can be chained without the need to assign to intermediate variables. the compiler is also ok to have the closure without an additional type hint. and the result inside the closure can be directly used as output, instead of unwrapping and wrapping it again.
```suggestion
    String::deserialize(deserializer)?
        .split(' ')
        .map(|document| DocumentId::new(document).map_err(de::Error::custom))
        .collect()
```